File: C++/문제_1654번.cpp
//문제 1654번 랜선 자르기 -> 이분 탐색
typedef long long ll;
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    cout.tie(NULL);
    int k, n;
    ll rope[10001];
    ll result = 0;
    cin >> k >> n;
    for(int i = 0;i<k;i++){
        cin >> rope[i];
    }
    ll left = 0;
    ll right = 20000000000000;
    while(left + 1 < right){
        ll mid = (left + right) / 2;
        int cnt = 0;
        for(int i = 0;i<k;i++){
            cnt+= rope[i] / mid;
        }
        if(cnt >= n){
            left = mid;
            result = max(result, left);
        }
        else if(cnt < n)
            right = mid;
    }
    cout << result;
}

File: JavaScript/문제_1654번.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
input[0] = input[0].split(' ');
let n = +input[0][0];
let m = +input[0][1];
input.shift();
let arr = [...input.map(o=>+o)];
let left = +0, right = +20000000000000;//길이 단위
let result = +0;
while (left + 1 < right) {
  let mid = Math.floor((left + right) / 2);//길이
  let sum = +0;//랜선의 개수
  for (let i = 0; i < arr.length; i++){
    sum += Math.floor(arr[i] / mid);
  }
  if (sum < m)
    right = mid;
  else{
    left = mid;
    result = Math.max(result, left);
  }
}
console.log(result);